File: antia_numerical_methods/f77/sub8/betcon1.f
!	To calculate the incomplete Beta function I_x(a,b) using
!	the continued fraction, called by BETAP
!
!	A,B : (input) Arguments for the complete Beta function
!	X : (input) Upper limit of integration defining the incomplete
!                       Beta function
!
!	Required routines : GAMMAL

      FUNCTION BETCON1(A,B,X)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION c(500),d(500)
      
51    FORMAT(' ** Roundoff error while evaluating the ',
     1  'continued fraction for incomplete Beta function at',/,'  A=',
     2  1PD12.4,'  B=',D12.4,'  X=',D12.4,'  CON. FRAC.=',D12.4)

      C(1)=1
      D(1)=1
      C(2)=1
      D(2)=1-X*(A+B)/(A+1)
      C1=C(2)/D(2)
      DO I=2,498,2
        M=I/2
        C(I+1)=C(I)+M*(B-M)*X*C(I-1)/((A+I-1)*(A+I))
        D(I+1)=D(I)+M*(B-M)*X*D(I-1)/((A+I-1)*(A+I))
        C(I+2)=C(I+1)-(A+M)*(A+B+M)*X*C(I)/((A+I+1)*(A+I))
        D(I+2)=D(I+1)-(A+M)*(A+B+M)*X*D(I)/((A+I+1)*(A+I))
!	Scale the numerator and denominator to prevent underflow/overflow
        IF(ABS(C(I+2)).GT.1.D200) THEN
          C(I+1)=C(I+1)/1.D200
          D(I+1)=D(I+1)/1.D200
          C(I+2)=C(I+2)/1.D200
          D(I+2)=D(I+2)/1.D200
        ENDIF
        IF(ABS(C(I+2)).LT.1.D-200) THEN
          C(I+1)=C(I+1)*1.D200
          D(I+1)=D(I+1)*1.D200
          C(I+2)=C(I+2)*1.D200
          D(I+2)=D(I+2)*1.D200
        ENDIF
        C2=C(I+2)/D(I+2)
        IF(ABS(C2-C1).LT.1.D-12) EXIT
        C1=C2
      ENDDO
      IF(C2.LT.0.0) PRINT 51,A,B,X,C2
      B1=A*LOG(X)+B*LOG(1-X)+LOG(C2)-LOG(A)
      B1=B1+GAMMAL(A+B)-GAMMAL(A)-GAMMAL(B)
      BETCON1=EXP(B1)
      END

File: antia_numerical_methods@Beliavsky/c/fun/betcon1.c
/*	To calculate the incomplete Beta function I_x(a,b) using
	the continued fraction, called by betap

	a,b : (input) Arguments for the complete Beta function
	x : (input) Upper limit of integration defining the incomplete
			Beta function

	Required routines : gammaln
*/


#include <math.h>
#include <stdio.h>

double gammaln(double x);

double betcon1(double a, double b, double x)
{
	double c[500],d[500],a1,b1,c1,c2;
	int i,m;

	c[0]=1; d[0]=1; c[1]=1; d[1]=1-x*(a+b)/(a+1);
	c2=c[1]/d[1]; c1=0.0; i=0;
	while(i<500 && fabs(c1-c2)>1.e-12)
	{
		i=i+2; m=i/2; c1=c2;
		c[i]=c[i-1]+m*(b-m)*x*c[i-2]/((a+i-1)*(a+i));
		d[i]=d[i-1]+m*(b-m)*x*d[i-2]/((a+i-1)*(a+i));
		c[i+1]=c[i]-(a+m)*(a+b+m)*x*c[i-1]/((a+i+1)*(a+i));
		d[i+1]=d[i]-(a+m)*(a+b+m)*x*d[i-1]/((a+i+1)*(a+i));
		if(fabs(c[i])>1.e200)
/*	Scale the numerator and denominator to prevent underflow/overflow */
		{
			c[i]=c[i]/1e200; c[i+1]=c[i+1]/1e200;
			d[i]=d[i]/1e200; d[i+1]=d[i+1]/1e200;
		}
		if(fabs(c[i])<1.e-200)
		{
			c[i]=c[i]*1e200; c[i+1]=c[i+1]*1e200;
			d[i]=d[i]*1e200; d[i+1]=d[i+1]*1e200;
		}
		c2=c[i+1]/d[i+1];
	}
	if(c2<0.0) printf(" ** Roundoff error while evaluating the continued fraction for incomplete Beta function at\n a= %e  b=%e  x= %e cont. frac.= %e\n",a,b,x,c2);
	b1=a*log(x)+b*log(1-x)+log(c2)-log(a);
	b1=b1+gammaln(a+b)-gammaln(a)-gammaln(b);
	return exp(b1);
}

